base/src/types.rs

```diff
@@ -610,24 +619,16 @@ impl<'a, Id, T> Iterator for ArgIterator<'a, T>
     type Item = &'a T;
     fn next(&mut self) -> Option<&'a T> {
         match **self.typ {
-            Type::Function(ref arg, ref return_type) => {
-                self.typ = return_type;
-                Some(&arg[0])
+            Type::App(ref app, ref args) if app.is_function() => {
```

Comment:
Maybe it is worth it to have `Type::as_function(&self) -> Option<(&T, &T)>` to make this a bit cleaner and less error prone?